src/librustdoc/html/static/js/search.js

```diff
@@ -133,34 +124,359 @@ window.initSearch = function(rawSearchIndex) {
-     * @return {{
-     *   in_args: Array<?>,
-     *   returned: Array<?>,
-     *   others: Array<?>
-     * }}
-     */
-    function execQuery(query, searchWords, filterCrates) {
-        function itemTypeFromName(typename) {
-            for (var i = 0, len = itemTypes.length; i < len; ++i) {
-                if (itemTypes[i] === typename) {
-                    return i;
+    function isWhitespace(c) {
+        return " \t\n\r".indexOf(c) !== -1;
+    }
+    function isSpecialStartCharacter(c) {
+        return "(<\"".indexOf(c) !== -1;
+    }
+    function isStopCharacter(c) {
+        return isWhitespace(c) || "),>-=".indexOf(c) !== -1;
+    }
+    function removeEmptyStringsFromArray(arr) {
+        for (var i = 0, len = arr.length; i < len; ++i) {
+            if (arr[i] === "") {
+                arr.splice(i, 1);
+                i -= 1;
+            }
+        }
+    }
+    function itemTypeFromName(typename) {
+        for (var i = 0, len = itemTypes.length; i < len; ++i) {
+            if (itemTypes[i] === typename) {
+                return i;
+            }
+        }
+        return NO_TYPE_FILTER;
+    }
+
+    function getStringElem(query, parserState, isInGenerics) {
+        if (isInGenerics) {
+            throw new Error("`\"` cannot be used in generics");
+        } else if (query.literalSearch) {
+            throw new Error("Cannot have more than one literal search element");
+        } else if (parserState.totalElems !== 0) {
+            throw new Error("Cannot use literal search when there is more than one element");
+        }
+        parserState.pos += 1;
+        while (parserState.pos < parserState.length && query.userQuery[parserState.pos] !== "\"") {
+            if (query.userQuery[parserState.pos] === "\\") {
+                // We ignore the next coming character.
+                parserState.pos += 1;
+            }
+            parserState.pos += 1;
+        }
+        if (parserState.pos >= parserState.length) {
+            throw new Error("Unclosed `\"`");
+        }
+        // To skip the quote at the end.
+        parserState.pos += 1;
+        query.literalSearch = true;
+    }
+    function skipWhitespaces(query, parserState) {
+        while (parserState.pos < parserState.length) {
+            var c = query.userQuery[parserState.pos];
+            if (!isWhitespace(c)) {
+                break;
+            }
+            parserState.pos += 1;
+        }
+    }
+    function skipStopCharacters(query, parserState) {
+        while (parserState.pos < parserState.length) {
+            var c = query.userQuery[parserState.pos];
+            if (!isStopCharacter(c)) {
+                break;
+            }
+            parserState.pos += 1;
+        }
+    }
+    function isPathStart(query, parserState) {
+        return parserState.pos + 1 < parserState.length &&
+            query.userQuery[parserState.pos] === ':' &&
+            query.userQuery[parserState.pos + 1] === ':';
```

Comment:
```suggestion
        return query.userQuery.slice(parserState.pos, parserState.pos + 2) == '::';
```

Same technique can be used below for `isReturnArrow`.

If `userQuery` is not long enough, that's okay - `slice` will return what's available.

Reply:
:+1: